### PR TITLE
[PEP 558 - WIP] bpo-30744: Trace hooks no longer reset closure state

### PR DESCRIPTION
**Pull request by ncoghlan**
(Declined as per the PEP deferral notice in https://github.com/python/peps/pull/3050/files - this reference implementation was useful at the time, but after the frame management changes in Python 3.11 any implementation of the updated semantics in PEP 558 or PEP 667 would be best off starting from scratch. The test cases from this branch might still be useful, but those can always be copied out to a new branch easily enough)

----
Previously, trace hooks running on a nested function
could incorrectly reset the state of a closure cell.

This avoids that by changing the semantics of the
f_locals namespace on function frames to use a
write-through proxy, such that changes are made
immediately rather than being written back some
arbitrary time later.

[PEP 558](https://www.python.org/dev/peps/pep-0558/) and [bpo-17960](https://bugs.python.org/issue17960) cover additional changes
to locals() and frame.f_locals semantics that are
part of this update.

Remaining TODO items before this PR could be considered complete:

- [ ] Resolve #27525 and resync this PR with the main branch
- [ ] Update PEP 558 to cover the performance improvements made to the proxy implementation (lazy initial cache refresh, share a single fast refs mapping between all proxy instances for a given frame)
- [x] Merge in Python 3.11 frame implementation changes
- [x] Fix fast locals proxy implementation to account for Python 3.11 frame implementation changes
- [x] Fix fast locals proxy implementation to follow the design changes in https://github.com/python/peps/pull/1787
- [x] Stop implicitly updating the local variable snapshot when running Python tracing functions
- [ ] Python API documentation updates for `locals()`, `vars()`, and potentially others (check all `locals` mentions)
- [ ] docstring updates for `locals()` (and potentially `vars()`)
- [ ] `PyEval_*` C API documentation updates
- [ ] `PyFrame_*` C API documentation updates
- [x] ~Fix C API header layout to follow modern conventions~ (#18052)
- [x] When rearranging headers, fix the clang warning/error about `PyFrameObject` redefinition
- [ ] Add refcount adjustment info for new C APIs
- [x] Migrate `eval()` default locals namespace to `PyLocals_Get()`
- [x] Migrate `exec()` default locals namespace to `PyLocals_Get()`
- [ ] Report an error if `IMPORT_STAR` opcode is encountered in a `CO_OPTIMIZED` frame
- [x] Add an explicit test for the new `PyFrame_LocalsToFast()` runtime error
- [x] Add explicit C API tests for `PyLocals_Get()` and `PyEval_GetLocals()` at different scopes
- [x] Implement and test the other new PyLocals and PyFrame functions
- [x] Implement and test the following methods on the fast locals proxy:
    - [x] `setdefault()`
    - [x] `popitem()`
    - [x] `clear()`
    - [x] `update()` (theoretically already implemented, but needs explicit tests)
- [ ] Add explicit tests for the behaviour of proxies that reference a cleared frame

Desirable code structure improvements (current plan for these is to put any duplicated into a sharable location in this PR, but file separate maintainability issues to migrate to using the shared code in the original locations, to avoid the diff size impact on this PR):

- [ ] Deduplicate the mutable mapping code copied from `odictobject.c`
- [x] ~Move the `DictProxy` code out of `descrobject.c`~ (obsolete task, as frame locals proxy is now independent of mapping proxy)

https://bugs.python.org/issue30744

### CODE REVIEW

File: Objects/frameobject.c [hunk 1]
```diff
@@ -1091,9 +1105,20 @@ fastlocalsproxy_write_to_frame(fastlocalsproxyobject *flp, PyObject *key, PyObje
...
         } else {
-            /* TODO: It's Python integer mapping into the fast locals array */
+            /* Fast ref is a Python int mapping into the fast locals array */
+            Py_ssize_t offset = PyLong_AsSsize_t(fast_ref);
+            Py_ssize_t max_offset = flp->frame->f_code->co_nlocals;
+            if (offset < 0) {
+                result = -1;
+            } else if (offset > max_offset) {
+                PyErr_Format(PyExc_SystemError,
+                             "Fast locals ref (%d) exceeds array bound (%d)",
```
Comment (ZackerySpytz) [inline Objects/frameobject.c]: `%zd` needs to be used with `Py_ssize_t` (these errors were fixed recently in d53fe5f407ff4b529628b01a1bcbf21a6aad5c3a / https://bugs.python.org/issue36254).

**Review by auvipy — CHANGES_REQUESTED** [review]
rebase needed and build failures needed addressing.

**Comment by ncoghlan** [comment]
@njsmith Aside from the TODO's in the PR itself, this is a pretty complete implementation of the current version of PEP 558 now.

**Comment by ncoghlan** [comment]
One other known issue: the header file updates still need to be modified to conform with the new conventions.

**Comment by scotchka** [comment]
I don't know if this is already addressed elsewhere, but I'm getting a strange error. Running the below script with python.exe built with this branch (I removed an unused parameter to make it compile):
```
def foo():
    import pdb; pdb.set_trace()
    x=1

foo()
```
gives
```
$ ./python.exe jump.py
> /Users/henry/pep558/cpython/jump.py(3)foo()
-> x=1
(Pdb) x=123
(Pdb) c
python.exe(69485,0x10961d5c0) malloc: *** error for object 0x1029ee790: pointer being freed was not allocated
python.exe(69485,0x10961d5c0) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```
but this error is absent for certain values of x, say:
```
$ ./python.exe jump.py
> /Users/henry/pep558/cpython/jump.py(3)foo()
-> x=1
(Pdb) x=1234
(Pdb) c
$
```

**Comment by ncoghlan** [comment]
Hmm, looks like I may have a refcounting bug, and the small integer cache is able to provoke it into crashing.

**Comment by scotchka** [comment]
> Hmm, looks like I may have a refcounting bug, and the small integer cache is able to provoke it into crashing.

I can report that not all small numbers trigger this error. On my machine `x = 11`, say, is fine. Most perplexing!

**Comment by ncoghlan** [comment]
Running `test_frame` consistently triggers a crash in `test_clear_locals` for me, so I'm focusing on that initially.

File: Objects/frameobject.c (hunk 1 above)
Comment (ncoghlan) [reply Objects/frameobject.c]: Fixed

**Comment by ncoghlan** [comment]
Last push resolves the test_frame segfault, but adds a TODO noting that the locals proxy may still misbehave after a frame has been cleared (or is in the process of being destroyed).

There's also a threading test which checks reference cycles aren't being created through the frames, which is currently failing (quite possibly for a related reason)

**Comment by ncoghlan** [comment]
Ugh, the refcyle issue detected by test_threading is kinda inherent in the current implementation: once `frame.f_locals` gets created, there's now a cyclic reference between the frame and its own locals proxy.

I'm going to try forcibly breaking the cycle when an optimized frame exits (before the eval loop drops its own reference).

**Comment by ncoghlan** [comment]
test_sys_settrace is segfaulting in the gc collection call in test_trash_stack, and it's not at all obvious to me why that's happening :(

@pitrou @gpshead Any tips on investigating that? (gdb wasn't very helpful, since it just told me that it was the gc.collect call blowing up)

I already tried going back to earlier versions of my branch, and wasn't able to work out exactly when I introduced it.

**Comment by ncoghlan** [comment]
https://github.com/ncoghlan/cpython/commit/7078632520279fd19ec59b8faec5ddd27ea4b64c fixed the test_sys_settrace crash (mismanaged refcounts when local variables were updated via the write-through proxy).

There's still another segfault in test_log_destroyed_pending_task in test_asyncio, though.

**Comment by ncoghlan** [comment]
It appears test_asyncio was hitting the same bug as test_frame (the pop/delete support in the write-through proxy wasn't quite right), as they're both passing for me after the latest change.

**Comment by ncoghlan** [comment]
This PR is now a solid proof of concept for the PEP, but there's a lot of cleanup to be done before it would be ready for merging (and that's assuming the PEP is accepted in its current form - the latest draft is still to be discussed on python-dev).

I've made notes in the PR description of all the missing pieces that I'm already aware of (including the Mac OS X build failure, which is a clang vs gcc issue).

**Comment by ncoghlan** [comment]
Latest round of PEP review resulted in several changes to the proposed C API, including a clearer separation between additions to the stable ABI and additions to the full CPython C API. I created #18052 as a preparatory refactoring that migrates frameobject.h to the modern header layout that clearly separates the two APIs.

**Comment by ncoghlan** [comment]
I made enough progress on migrating to the redesigned API described in https://discuss.python.org/t/pep-558-defined-semantics-for-locals/2936/11?u=ncoghlan that next weekend I'll update the PEP to cover this version of the API (I didn't want to do that until I'd made the initial draft updates here).

**Comment by ncoghlan** [comment]
Thanks to a suggestion by @markshannon, I was able to get rid of the circular reference between the frame and the fast locals proxy objects (the frame has gone back to holding only the dynamic snapshot that it stores now, and the proxies are created on demand as part of the "frame.f_locals" get descriptor)

**Comment by ncoghlan** [comment]
I merged in the recent Python 3.11 frame implementation updates, and got everything building again, but the fast locals proxy and locals() snapshot building for optimised frames aren't working right yet (the deconflicted code is populating the locals snapshot when it should be building the fast locals proxy reference map, and it isn't otherwise populating the locals snapshot at all)

**Comment by ncoghlan** [comment]
@markshannon @ericsnowcurrently I've hit a problem with MAKE_CELL and the fast locals proxy that I think you may be able to provide some advice on.

My current plan is as follows:
* ordinary locals go into the fast reference map with integer offset values
* free variables go into the fast reference map with cell values
* cell variables go into the fast reference map with cell values if MAKE_CELL has already executed
* cell variables go into the fast reference map with integer offset values if MAKE_CELL has NOT executed
* the fast locals proxy access methods check the variable kind when an integer offset is stored in the fast reference map and the stored fast locals object is a cell. If the variable is a cell variable, and MAKE_CELL has executed, then the proxy fixes itself up to start using the cell as normal. If MAKE_CELL still hasn't executed yet, then it manipulates that field as if it were an ordinary local variable.

In effect, cell variables are treated like local variables until MAKE_CELL executes, and then like free variables afterwards.

**Comment by ncoghlan** [comment]
Implementation is now largely in line with the latest PEP version, but there's a refcounting bug to hunt down somewhere.

**Comment by ncoghlan** [comment]
`test_pdb` fails, as `pdb` writes `__return__` and `__exception__` attributes back to arbitrary frames. I'm going to fix the proxy to pass unknown keys through to `f_locals` rather than failing outright.

**Comment by ncoghlan** [comment]
All test issues related to the PEP implementation update are resolved.

https://github.com/python/peps/pull/2030/files updates the PEP to align with the revised implementation. The biggest point is having to bring back support for storing arbitrary values in order to avoid breaking `pdb`.

The ABI check complaining is legit - I haven't implemented all the new C API functions yet

**Comment by ncoghlan** [comment]
I think I came up with a resolution to the frame proxy performance issue that is sufficient without being overly complicated: the proxy operations with an expected algorithmic complexity that is less than O(n) simply assume the snapshot is already up to date. The operations that are O(n) anyway still ensure the snapshot is up to date as part of their execution.

If the assumption is ever wrong, there's now an explicit sync() method on the proxy to make it true.

**Comment by ncoghlan** [comment]
Pushed a few changes that reduce the likelihood of the f_locals cache desynchronising from the frame locals:
* every get/set operation on the frame variables refreshes the corresponding f_locals cache entry
* containment checks run a full item retrieval internally (so they update the cache, but also correctly return false for variable names that are currently unbound)

I haven't incorporated those latest tweaks into the PEP text yet (I'll do that before publishing the latest version next weekend).

**Comment by ncoghlan** [comment]
I started fleshing out the test suite a bit more to see if the implicit cache refresh logic felt right. The only thing that felt off was equality tests implying a cache refresh - those can be super quick when they're negative, so always incurring an O(n) cache refresh for a check where the answer could potentially be determined by an O(1) length check didn't seem right. So equality tests joined the list of operations that assume the cache is already up to date.

Those dict API tests aren't finished yet, so they emit a forced failure at the end.

New tests that *are* finished are the ones for the full public C API described in the PEP. These are mostly straightforward ctypes tests, although it did take me a few attempts to get the refcounting right for the APIs that return borrowed references (I'm still not 100% sure it is correct, but if it's still wrong it is now erring on the side of counting high rather than low- the initial tests segfaulted due to the locals cache getting deleted early).

**Comment by ncoghlan** [comment]
@njsmith I've submitted the latest iteration of the PEP for your review and pronouncement: https://mail.python.org/archives/list/python-dev@python.org/thread/RR62MUNEQPH2HLI6YE4VOQH6JQT64LPO/

**Comment by ericsnowcurrently** [comment]
(Sorry, I was AFK for a couple weeks.)

> @markshannon @ericsnowcurrently I've hit a problem with MAKE_CELL and the fast locals proxy that I think you may be able to provide some advice on.

Just to be sure we're talking about the same thing, this is about the proxy exposed as the Python-level readonly `f_locals` attribute on `PyFrameObject` objects.  If I misunderstood then please clarify.

As far as I understood the PEP, `PyFrame_FastToLocalsWithError()` and `PyFrame_LocalsToFast()` should be the template for the behavior.  That seems to be what you've done. 😄

> 
> My current plan is as follows:
> 
> * ordinary locals go into the fast reference map with integer offset values

correct

> * free variables go into the fast reference map with cell values

correct

(They aren't exposed by the proxy though, right?)

> * cell variables go into the fast reference map with cell values if MAKE_CELL has already executed

correct

> * cell variables go into the fast reference map with integer offset values if MAKE_CELL has NOT executed

This doesn't sound right but your explanation below is correct, so I probably misunderstood what you said here.

Just to be clear: if `MAKE_CELL` hasn't run yet (but will) then the "fast" entry should be set directly to the value.  Creation of the enclosing cell should be left exclusively to `MAKE_CELL`.

> * the fast locals proxy access methods check the variable kind when an integer offset is stored in the fast reference map and the stored fast locals object is a cell. If the variable is a cell variable, and MAKE_CELL has executed, then the proxy fixes itself up to start using the cell as normal. If MAKE_CELL still hasn't executed yet, then it manipulates that field as if it were an ordinary local variable.
> 
> In effect, cell variables are treated like local variables until MAKE_CELL executes, and then like free variables afterwards.

Yeah, this is correct.  So I think I misunderstood you point above about "if MAKE_CELL has NOT executed".

FWIW, here's what I wrote above before I read your last bullet 😄 :

> In `_PyEval_MakeFrameVector()` the "fast" entry for cell variable's offset is set to either null, a cell object from `func.__closure__` (for free vars), or an object supplied as an argument (or default argument).  In cases where it is a cell var, `MAKE_CELL` will run ASAP, at which point the "fast" entry will then hold the cell object containing whatever value was populated by `_PyEval_MakeFrameVector()`, or (with PEP 558) the proxy (or possibly even an object set via `PyFrame_LocalsToFast()`).
> 
> So just as happens now in `PyFrame_LocalsToFast()`, if `MAKE_CELL` hasn't run yet then the "fast" entry should be set to the assigned object (or NULL for deletion) rather than setting it to a new cell holding that value.
> 
> The alternative would be to add a branch to `MAKE_CELL` to determine if the cell was already set.  However that branch is expensive/complicated because we can't just check if the existing value is a cell--a cell object could be passed as an argument. ☹️  I originally took this approach and ran into that problem.
> 
> The simpler solution was to always leave the job of creating the cell to `MAKE_CELL`.  The only catch is that `PyFrame_FastToLocalsWithError()` and `PyFrame_LocalsToFast()` (and presumably the new proxy) must check if `MAKE_CELL` has already run.  (See `_PyFrame_OpAlreadyRan()`.)  Otherwise if they find the "fast" entry value is a cell then they wouldn't know if they need to replace it or replace its value.
> 
> So, the proxy should do what `PyFrame_LocalsToFast()` does: set the value as-is (if `MAKE_CELL` hasn't run yet) or update the cell (if it has run).

**Comment by ericsnowcurrently** [comment]
In thinking about the proxy, I had some questions:

* What happens with new variables that get added to the locals (e.g. while debugging)?
   * Presumably they are stored on the proxy itself.  There certainly doesn't exist any "fast" entry on the frame for such a new variable.
   * Other proxies of the same locals will not have these new locals, right?
* Do free vars still get exposed by the proxy even if they aren't used in the current scope?
   * Presumably the answer is yes since that is the behavior of `locals()` currently.
   * FWIW, I expected the opposite behavior from `locals()`: that it exposes only variables bound in the scope.  Unused free vars are only tied to the frame for the sake of building an inner function's `__closure__`.  So I was surprised when I ran a quick experiment and found that those otherwise invisible variables are exposed by `locals()`.  Other unused non-locals certainly aren't exposed by `locals()`, so it was a bit unintuitive for me.  Unfortunately, I expect that ship has long since sailed. ☹️ 
* Likewise for implicit cell vars, like `__class__` (for `super()`)?
   * This surprised me even more.  I think of `__class__` as even more invisible than those hidden variables needed for `MAKE_FUNCTION`.  I'm sure it's likewise too late to change this. ☹️

**Comment by ncoghlan** [comment]
I've pushed an update that resolves almost all of my own "PEP 558 TODO" items in the code (and ticked them off in the checklist above as well):
* the proxy implements the full mutable mapping API (and this support is tested)
* the proxy now holds off on refreshing the value cache until the first operation that assumes the value cache is up to date (so if you only access individual keys by name, you'll never trigger a full cache refresh, allowing O(1) tracing operation regardless of the number of variables on the frame)
* the fast refs mapping is now stored on the frame object, so 

I will be adding another todo item to the checklist, which is to explicitly test the handling of cleared frames. Implementing `proxy.clear()` exposed some inconsistencies between the established `FastToLocals()` logic and the proxy's behaviour when accessing cell variables via a cleared frame, but no unit tests failed when I changed the behaviour to be consistent with the `FastToLocals()` precedent.

For the merge conflict, I'm hoping to get https://github.com/python/cpython/pull/27525/ merged before I try to sync this branch with the main branch again. It was my first attempt at doing that which prompted me to file bpo-44800 in the first place, as it was quite unclear how the conflicts should be resolved given the currently checked in variable naming conventions.

**Comment by ncoghlan** [comment]
@ericsnowcurrently:

* I think what the implementation is doing for "not yet created" cell objects aligns with what you say it should be doing. The fast refs entry for a "not yet created" cell object is like the entry for a local variable: a Python integer giving the location of the cell in the fast locals array. It is treated as an unbound name while it is in that state. However, each time it is accessed for any reason, the proxy checks to see if the cell object has been created by `MAKE_CELL` since the last time the proxy checked. If it has, it fixes up the fast refs entry to refer directly to that cell object instead of to the fast locals array slot. It's still only `MAKE_CELL` that will ever create the cell object, though - the proxy just promotes it to the fast refs mapping once it notices it has been created.
* any extra variables that get set are still stored on the underlying frame rather than each proxy object having its own separate storage location. They're just not stored in the fast locals array - they're stored in the same locals dictionary that they're stored in now (which provides backwards compatibility and interoperability with the `PyEval_GetLocals()` API). That fast locals dictionary is *also* used to cache the name->value mapping for the fast locals, both for `PyEval_GetLocals()` compatibility, and to avoid having to make routine operations on the proxy (like checking how many names are currently bound to values) consistently `O(n)` (instead the proxy refreshes the cache the first time it needs to use it, and then after that assumes it is still up to date - since proxies are cheap to create, the easiest way to refresh them is to throw the old proxy away and make a new one, but there's also an explicit cache sync method defined)
* yes, the proxy implementation still exposes all the surprisingly visible cell references that `LocalsToFast()` does, and it makes attempts to write to them actually work. As you say, doing anything else would almost certainly create a compatibility problem at this point. It does mean that calling `clear()` on the proxy can have more side effects than calling `clear()` on the frame, but that's consistent with the way `PyFrame_LocalsToFast()` has historically worked in tracing functions (deleting a key from `f_locals` would unbind it when returning from the tracing function, regardless of whether it was a local variable or a cell reference)

**Comment by mentalisttraceur** [comment]
I have a concern about raising `RuntimeError` from `PyFrame_LocalsToFast`.

Was it possible for `PyFrame_LocalsToFast` before this change to ever raise `RuntimeError` for any other reason?

If not, then it would be nice to have that explicitly documented (at least in the PEP, maybe elsewhere if appropriate and not too noisy to include it).

But `RuntimeError` is raised for internal problems, like if memory allocations fail, right? So it seems like the kind of error type that's liable to fly out of any internal function, with no real guarantees that it can't?

So I think it might be a bit of a challenge for third parties to write code that

1. work on Python versions before this change,
2. will work on later Python version after this change, and
3. won't ever *hide any other errors on versions before this change*, and
4. does all that reasonably cleanly.

Because it seems like code released today:

1. can't start catching just any `RuntimeError` from `PyFrame_LocalsToFast`,
2. doesn't have guarantees about exactly what subclass or error message of `RuntimeError` to expect based on this PEP, and
3. can't just check `sys.version_info` because it seems like we don't have certainty about when this PEP will make it in.

For what it's worth, this isn't *insurmountable* - in [my little toy library that mutates functions' variables through `f_locals`](https://pypi.org/project/dynamicscope), I settled on *doing a function call and trying to modify a variable in its frame, checking if it worked, and only if it didn't trying to import the relevant stuff if available for that Python*. I think this is probably the most ideal and robust *in terms of portability* across all versions of all implementations of Python where mutating a frame's locals is possible, and elegantly avoids ever causing the `RuntimeError` that this PEP adds on any implementation+version of Python that implements this PEP. But it's [it takes all this code](https://github.com/mentalisttraceur/python-dynamicscope/blob/17ff81972f696fcf05cb7eaaf4e311b32520ae39/dynamicscope.py#L64-L97), might be fragile in other ways or have other downsides I haven't realized yet, and wasn't trivial to think of.

**TL;DR**: this change takes `PyFrame_LocalsToFast` from *necessary* to *unusable*, so it would be nice if the error should be programmatically easy to distinguish for the code that was using it while it was necessary.

**Comment by ncoghlan** [comment]
Closing as per the deferral notice in https://github.com/python/peps/pull/3050/files
